File: Semestrul 4/Medii de proiectare si programare/Laboratoare/FirmaTransportC#/Network/protocol/TransportServicesProxy.cs
using FirmaTransportC_.Model;
using FirmaTransportC_.Networking.dto;
using FirmaTransportC_.Services;
using log4net;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;

namespace FirmaTransportC_.Networking.protocol
{
    public class TransportServicesProxy : ITransportService
    {
        private string host;
        private int port;

        private IObserver client;
        private NetworkStream stream;
        private TcpClient connection;
        private Queue<Response> responses;
        private volatile bool finished;
        private EventWaitHandle _waitHandle;
        private static readonly ILog log = LogManager.GetLogger(typeof(TransportServicesProxy));

        public TransportServicesProxy(string host, int port)
        {
            this.host = host;
            this.port = port;
            responses = new Queue<Response>();
        }

        public void Login(User user, IObserver client)
        {
            initializeConnection();
            sendRequest(JsonProtocolUtils.CreateLoginRequest(user));
            Response response = readResponse();
            if(response.Type == ResponseType.OK)
            {
                this.client = client;
                log.Info("Login successful");
            }

            if(response.Type == ResponseType.ERROR)
            {
                string errorMessage = response.ErrorMessage ?? "Unknown error";
                closeConnection();
                throw new ServiceException("Login failed: " + errorMessage);
            }
        }

        public void Logout(User user)
        {
            sendRequest(JsonProtocolUtils.CreateLogoutRequest(user));
            Response response = readResponse();
            closeConnection();
            if (response.Type == ResponseType.ERROR)
            {
                throw new ServiceException(response.ErrorMessage);
            }
        }

        public List<Trip> GetAllTrips()
        {
            sendRequest(JsonProtocolUtils.CreateGetTripsRequest());
            Response response = readResponse();
            if(response.Type != ResponseType.TRIPS_LIST)
            {
                throw new ServiceException("Unexpected response type: " + response.Type);
            }

            return DTOUtils.GetFromDTO(response.Trips);
        }

        public List<Seat> GetSeatsForTrip(Trip trip)
        {
            sendRequest(JsonProtocolUtils.CreateGetSeatsRequest(trip));
            Response response = readResponse();
            if(response.Type != ResponseType.SEATS_LIST)
            {
                throw new ServiceException("Unexpected response type: " + response.Type);
            }

            return DTOUtils.GetFromDTO(response.Seats, trip.Id);
        }

        public List<Trip> SearchTrips(Trip trip)
        {
            sendRequest(JsonProtocolUtils.CreateSearchTripsRequest(DTOUtils.GetSearchDTO(trip)));
            Response response = readResponse();
            if(response.Type != ResponseType.SEARCH_RESULT)
            {
                throw new ServiceException("Unexpected response type: " + response.Type);
            }

            return DTOUtils.GetFromDTO(response.Trips);
        }

        public void MakeReservation(Reservation reservation)
        {
            sendRequest(JsonProtocolUtils.CreateMakeReservationRequest(reservation));
            Response response = readResponse();
            if(response.Type != ResponseType.OK)
            {
                throw new ServiceException("Unexpected response type: " + response.Type);
            }
        }

        public void CancelReservation(Reservation reservation)
        {
            sendRequest(JsonProtocolUtils.CreateCancelReservationRequest(reservation));
            Response response = readResponse();
            if(response.Type != ResponseType.OK)
            {
                throw new ServiceException("Unexpected response type: " + response.Type);
            }
        }

        private void initializeConnection()
        {
            try
            {
                connection = new TcpClient(host, port);
                stream = connection.GetStream();
                finished = false;
                _waitHandle = new AutoResetEvent(false);
                startReader();
            }
            catch (Exception e)
            {
                log.Error(e);
                throw new ServiceException("Error connecting to server: " + e.Message);
            }
        }

        private void closeConnection()
        {
            finished = true;
            try
            {
                stream.Close();
                connection.Close();
                _waitHandle.Close();
                client = null;
            }
            catch (Exception e)
            {
                log.Error(e);
            }
            finally
            {
                stream = null;
                connection = null;
                _waitHandle = null;
                client = null;
            }
        }

        private void sendRequest(Request request)
        {
            try
            {
                lock (stream)
                {

                    string jsonRequest = JsonSerializer.Serialize(request);
                    log.DebugFormat("Sending request {0}", jsonRequest);
                    byte[] data = Encoding.UTF8.GetBytes(jsonRequest + "\n");
                    stream.Write(data, 0, data.Length);
                    stream.Flush();

                }
            }
            catch (Exception e)
            {
                throw new ServiceException("Error sending request " + e.Message);
            }

        }

        private Response readResponse()
        {
            Response response = null;
            try
            {
                _waitHandle.WaitOne();
                lock (responses)
                {
                    response = responses.Dequeue();

                }

                if(response.ErrorMessage != null)
                {
                    throw new ServiceException("Error response: " + response.ErrorMessage);
                }

                return response;
            }
            catch (Exception e)
            {
                throw new ServiceException("Error reading response " + e.Message);
            }
        }

        private void startReader()
        {
            Thread tw = new Thread(run);
            tw.IsBackground = true;
            tw.Start();
        }

        private void handleUpdate(Response response)
        {
            log.Debug("handling update " + response);
            if (response.Type != ResponseType.UPDATE || client == null)
            {
                return;
            }

            try
            {
                List<Trip> trips = DTOUtils.GetFromDTO(response.Trips);
                client.tripsUpdated(trips);
            }
            catch (Exception e)
            {
                log.Error("Error handling update " + e);
            }
        }   

        private bool isUpdate(Response response)
        {
            return response.Type == ResponseType.UPDATE;
        }

        public virtual void run()
        {
            using StreamReader reader = new StreamReader(stream, Encoding.UTF8);
            while (!finished)
            {
                try
                {
                    string responseJson = reader.ReadLine();
                    if (string.IsNullOrEmpty(responseJson))
                        continue;

                    Response response = JsonSerializer.Deserialize<Response>(responseJson);
                    log.Debug("response received " + response);

                    if (isUpdate(response))
                    {
                        handleUpdate(response);
                    }
                    else
                    {
                        lock (responses)
                        {
                            responses.Enqueue(response);
                        }
                        _waitHandle.Set();
                    }
                }
                catch (IOException) when (finished)
                {
                    break;
                }
                catch (ObjectDisposedException) when (finished)
                {
                    break;
                }
                catch (Exception e)
                {
                    if (!finished)
                    {
                        log.Error("Reading error " + e);
                    }
                    break;
                }
            }
        }

    }
}

File: Semestrul 4/Medii de proiectare si programare/Laboratoare/FirmaTransportJava/Networking/src/main/java/transport/networking/protocol/TransportServicesProxy.java
package transport.networking.protocol;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import transport.model.Reservation;
import transport.model.Seat;
import transport.model.Trip;
import transport.model.User;
import transport.networking.dto.DTOUtils;
import transport.networking.dto.ReservationDTO;
import transport.networking.dto.SearchDTO;
import transport.networking.dto.TripDTO;
import transport.networking.dto.UserDTO;
import transport.services.IObserver;
import transport.services.ITransportService;
import transport.services.ServiceException;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class TransportServicesProxy implements ITransportService {

    private final String host;
    private final int port;
    private static final Logger logger = LogManager.getLogger(TransportServicesProxy.class);

    private IObserver client;
    private ObjectInputStream input;
    private ObjectOutputStream output;
    private Socket connection;

    private final BlockingQueue<Response> qresponses;
    private volatile boolean finished;

    public TransportServicesProxy(String host, int port) {
        this.host = host;
        this.port = port;
        this.qresponses = new LinkedBlockingQueue<>();
    }

    private void initializeConnection() throws ServiceException {
        try {
            connection = new Socket(host, port);
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            finished = false;
            startReader();
        } catch (IOException e) {
            throw new ServiceException("Error initializing connection: " + e.getMessage());
        }
    }

    private void closeConnection() {
        finished = true;
        try {
            if (input != null) {
                input.close();
            }
            if (output != null) {
                output.close();
            }
            if (connection != null) {
                connection.close();
            }
        } catch (IOException e) {
            logger.error("Error closing connection", e);
        } finally {
            client = null;
            input = null;
            output = null;
            connection = null;
        }
    }

    private void sendRequest(Request request) throws ServiceException {
        if (output == null) {
            throw new ServiceException("Not connected to server");
        }
        try {
            output.writeObject(request);
            output.flush();
        } catch (IOException e) {
            throw new ServiceException("Error sending request: " + e.getMessage());
        }
    }

    private Response readResponse() throws ServiceException {
        try {
            Response response = qresponses.take();
            if (response.getType() == ResponseType.ERROR) {
                throw new ServiceException(String.valueOf(response.getData()));
            }
            return response;
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new ServiceException("Interrupted while waiting for server response");
        }
    }

    private void startReader() {
        Thread thread = new Thread(new ReaderThread());
        thread.setDaemon(true);
        thread.start();
    }

    @SuppressWarnings("unchecked")
    private void handleUpdate(Response response) {
        if (response.getType() != ResponseType.UPDATE || client == null) {
            return;
        }

        try {
            List<Trip> trips = (List<Trip>) response.getData();
            client.tripsUpdated(trips);
        } catch (Exception e) {
            logger.error("Error dispatching update to observer", e);
        }
    }

    @Override
    public void login(User user, IObserver client) throws ServiceException {
        initializeConnection();

        UserDTO userDTO = DTOUtils.getDTO(user);
        sendRequest(new Request(RequestType.LOGIN, userDTO));

        Response response = readResponse();
        if (response.getType() != ResponseType.OK) {
            closeConnection();
            throw new ServiceException("Unexpected response type on login: " + response.getType());
        }

        this.client = client;
    }

    @Override
    public void logout(User user) throws ServiceException {
        try {
            UserDTO userDTO = DTOUtils.getDTO(user);
            sendRequest(new Request(RequestType.LOGOUT, userDTO));
            readResponse();
        } finally {
            closeConnection();
        }
    }

    @Override
    @SuppressWarnings("unchecked")
    public List<Trip> getAllTrips() throws ServiceException {
        sendRequest(new Request(RequestType.GET_TRIPS, null));
        Response response = readResponse();

        if (response.getType() != ResponseType.TRIPS_LIST) {
            throw new ServiceException("Unexpected response type for trips: " + response.getType());
        }

        return (List<Trip>) response.getData();
    }

    @Override
    @SuppressWarnings("unchecked")
    public List<Trip> searchTrips(Trip trip) throws ServiceException {
        SearchDTO search = DTOUtils.getSearchDTO(trip);
        sendRequest(new Request(RequestType.SEARCH_TRIPS, search));
        Response response = readResponse();

        if (response.getType() != ResponseType.SEARCH_RESULT) {
            throw new ServiceException("Unexpected response type for search: " + response.getType());
        }

        return (List<Trip>) response.getData();
    }

    @Override
    @SuppressWarnings("unchecked")
    public List<Seat> getSeatsForTrip(Trip trip) throws ServiceException {
        TripDTO tripDTO = DTOUtils.getTripDTO(trip);
        sendRequest(new Request(RequestType.GET_SEATS_FOR_TRIP, tripDTO));
        Response response = readResponse();

        if (response.getType() != ResponseType.SEATS_LIST) {
            throw new ServiceException("Unexpected response type for seats: " + response.getType());
        }

        return (List<Seat>) response.getData();
    }

    @Override
    public void makeReservation(Reservation reservation) throws ServiceException {
        ReservationDTO reservationDTO = DTOUtils.getDTO(reservation);
        sendRequest(new Request(RequestType.MAKE_RESERVATION, reservationDTO));

        Response response = readResponse();
        if (response.getType() != ResponseType.OK) {
            throw new ServiceException("Unexpected response type for reservation: " + response.getType());
        }
    }

    @Override
    public void cancelReservation(Reservation reservation) throws ServiceException {
        ReservationDTO reservationDTO = DTOUtils.getDTO(reservation);
        sendRequest(new Request(RequestType.CANCEL_RESERVATION, reservationDTO));

        Response response = readResponse();
        if (response.getType() != ResponseType.OK) {
            throw new ServiceException("Unexpected response type for cancel: " + response.getType());
        }
    }

    private boolean isUpdate(Response response) {
        return response.getType() == ResponseType.UPDATE;
    }

    private class ReaderThread implements Runnable {
        @Override
        public void run() {
            while (!finished) {
                try {
                    Object response = input.readObject();
                    Response castResponse = (Response) response;
                    if (isUpdate(castResponse)) {
                        handleUpdate(castResponse);
                    } else {
                        qresponses.put(castResponse);
                    }
                } catch (IOException | ClassNotFoundException e) {
                    if (!finished) {
                        logger.error("Reader error", e);
                    }
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    logger.error("Reader interrupted", e);
                    return;
                }
            }
        }
    }
}